examples/library/genetic_algorithm_SNAP/libmod_optimize.py

```diff
@@ -0,0 +1,566 @@
+    nf = ne
+    # total
+    nh = ne + nf
+
+    # weights for energy and force rmse in the optimizer cost function
+    etot_weight = 1.0
+    ftot_weight = 5.0
+    # allowed scaling factors for energy weights
+    my_w_ranges = [1.e-3,1.e-2,1.e-1,1,1.e1,1.e2,1.e3]
+    eranges = [my_w_ranges]
+    # allowed scaling factors for force weights
+    my_ef_ratios = [0.1,1,10,100]
+    ffactors = [my_ef_ratios]
+
+    # selection method (only tournament is currently implemented)
+    selection_method = 'tournament'
+
+    # cross over (parenting) and mutation hyperparameters
+    r_cross = 0.9
+    r_mut = 0.1
+
+    # convergence threshold for full function (value of RMSE E + RMSE F at which simulation is terminated" 
+    convthr = 0.005 
+    # fraction of ngenerations to start checking for convergence (convergence checks wont be performed very early)
+    conv_check = 1.2
+
+    #End optimization hyperparameters
+    #---------------------------------------------------------------------------
+
+    # set up generation 0
+    best_eval = 9999999.9999
+    conv_flag = False
+    first_seeds = seedsi[:population_size+1]
+    hp = HyperparameterStruct(ne,nf,eranges,ffactors)
+    population = [hp.random_params(inputseed=first_seeds[ip]) for ip in range(population_size)]
+    generation = 0
+
+    best_evals = [best_eval]
+    best = tuple(population[0])
+    sim_seeds = seedsi[population_size:]
+    np.random.seed(sim_seeds[generation])
+    w_combo_delta = np.ones(len(gtks))
+    # delta function to zero out force weights on structures without forces
+    ef_rat_delta = np.array([1.0 if 'Volume' not in gti else 0.0 for gti in gtks])
+    while generation <= ngenerations and best_eval > convthr and not conv_flag:
+        scores = []
+        # current generation
+        for creature in population:
+            creature_ew, creature_ffac = tuple(creature.reshape(2,ne).tolist())
+            creature_ew = tuple(creature_ew)
+            creature_ffac = tuple(creature_ffac)
+            update_weights(creature_ew, creature_ffac, gtks = gtks, size_b = size_b, grouptype = grouptype)
+            costi = fit_and_cost(snap,[etot_weight,ftot_weight])
+            scores.append(costi)
+            #NOTE to add another contribution to the cost function , you need to evaluate it in the loop
+            # and add it to the fit_and_cost function
+            # if this involves a lammps simulation, you will have to print potentials at the different steps
+            # to run the lammps/pylammps simulation. To do so, the fitsnap output file name prefix should
+            # be updated per step, then snap.write_output() should be called per step. This will likely increase
+            # the optimization time.
+        # Anything printed with pt.single_print will be included in output file.
+        pt.single_print('generation,scores,popsize:',generation,len(scores),population_size)
+        # Print generation and best fit.
+        bestfit = min(scores)
```

Comment:
Added this as a suggestion - is this the best way to track progress?
There's a lot of screen output so I dumped that into a file with the `screen2file` command line option.
We can use normal `print` to only output important info to the screen.